### PR TITLE
Refactor resource group slots

### PR DESCRIPTION
**Pull request by xiong-gang**
1) change slot pool to a list linked with pointer
2) refactor get/put/aquire/release slot.
3) change DDL callback functions

### CODE REVIEW

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -1234,13 +1237,33 @@ getSlot(ResGroupData *group)
...
+	/* Now actually get a free slot */
+	slot = slotpoolAllocSlot();
+
+	group->nRunning++;
+
+	return slot;
+}
+
+static bool
+groupReserveMemQuota(ResGroupData *group)
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: As we now use this helper function to reserver mem quota, I think it's better to also have one to release mem quota.

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -1308,27 +1316,20 @@ putSlot(void)
...
- * Acquire a resource group slot
- *
- * Call this function at the start of the transaction.
- * This function set current resource group in MyResGroupSharedInfo,
- * and current slot id in MyProc->resSlotId.
- */
-static void
-ResGroupSlotAcquire(void)
+static ResGroupData *
+decideResGroup(void)
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: comments please.

File: src/backend/utils/resgroup/resgroup.c [hunk 3]
```diff
@@ -1344,37 +1345,58 @@ ResGroupSlotAcquire(void)
...
 	}
 
-	/*
-	 * it's neccessary to set group to self before we
-	 * got signal and goes to ResGroupWaitCancel
-	 */
 	selfSetGroup(group);
+	LWLockRelease(ResGroupLock);
+	return group;
+}
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: blank line please.

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: also need a forward function declaration.

File: src/backend/utils/resgroup/resgroup.c [hunk 4]
```diff
@@ -1980,17 +1999,21 @@ AssignResGroupOnMaster(void)
...
 	PG_TRY();
 	{
-		/* Acquire slot */
-		Assert(pResGroupControl != NULL);
-		Assert(pResGroupControl->segmentsOnMaster > 0);
-		Assert(selfIsUnassigned());
-		ResGroupSlotAcquire();
+retry:
+		group = decideResGroup();
 		Assert(selfIsAssignedValidGroup());
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: I always have a question here: after we have decided the group and released the lwlock, will it then be dropped concurrently before we run to above assertion or operations?

File: src/backend/utils/resgroup/resgroup.c [hunk 5]
```diff
@@ -2462,8 +2477,6 @@ selfValidateResGroupInfo(void)
 	Assert(self->memUsage >= 0);
 
 	AssertImply(self->groupId != InvalidOid,
-				self->slotId != InvalidSlotId);
-	AssertImply(self->groupId != InvalidOid,
 				self->group != NULL);
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: As we no longer require the group and slot to be both set or unset, we should update the function comments now.

**Comment by pivotal-ning-yu** [comment]
In general LGTM.

File: src/backend/utils/resgroup/resgroup.c [hunk 6]
```diff
@@ -2049,25 +2072,21 @@ UnassignResGroup(void)
 	if (Gp_role == GP_ROLE_DISPATCH)
 	{
 		/* Release the slot */
-		ResGroupSlotRelease();
+		groupReleaseSlot(group, slot);
```
Comment (pengzhout) [inline src/backend/utils/resgroup/resgroup.c]: If there is a entry-db, is it safe to release group slot without checking the nprocs?

**Comment by pivotal-ning-yu** [comment]
As @xiong-gang is on vacation this week I'll take care of this PR, but as I have no permission to push to his repo I'll open a PR instead.

File: src/backend/utils/resgroup/resgroup.c [hunk 7]
```diff
@@ -1131,8 +1122,9 @@ static void
 selfDetachSlot(ResGroupData *group, ResGroupSlotData *slot)
 {
 	groupDecMemUsage(group, slot, self->memUsage);
-	slot->nProcs--;
+	pg_atomic_sub_fetch_u32((pg_atomic_uint32*) &slot->nProcs, 1);
 	AssertImply(slot->nProcs == 0, slot->memUsage == 0);
+	selfUnsetSlot();
```
[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Moved the call to `selfUnsetSlot()` into `selfAttachToSlot()` in https://github.com/greenplum-db/gpdb/pull/3474

File: src/backend/utils/resgroup/resgroup.c [hunk 8]
```diff
@@ -1234,13 +1237,33 @@ getSlot(ResGroupData *group)
...
 	if (group->nRunning >= caps->concurrency.proposed)
-		return InvalidSlotId;
+		return NULL;
 
-	groupAcquireMemQuota(group, caps);
+	if (!groupReserveMemQuota(group))
+		return NULL;
+
+	/* Now actually get a free slot */
+	slot = slotpoolAllocSlot();
```
[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Assertion added in https://github.com/greenplum-db/gpdb/pull/3474

File: src/backend/utils/resgroup/resgroup.c (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: `groupReleaseMemQuota()` added in https://github.com/greenplum-db/gpdb/pull/3474

BTW, there was a pair of `groupReleaseMemQuota()` and `groupAcquireMemQuota()` to operate memory quota in the global memory pool, they are now renamed to `mempoolReleaseMemQuota()` and `mempoolReserveMemQuota()`

File: src/backend/utils/resgroup/resgroup.c [hunk 9]
```diff
@@ -1256,43 +1279,28 @@ getSlot(ResGroupData *group)
...
-	/* Now actually get a free slot */
-	slotId = slotPoolAlloc();
-	Assert(slotId != InvalidSlotId);
-
-	group->nRunning++;
-
-	return slotId;
+	return true;
 }
-
```
[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Updated in https://github.com/greenplum-db/gpdb/pull/3474

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Updated in https://github.com/greenplum-db/gpdb/pull/3474

File: src/backend/utils/resgroup/resgroup.c (hunk 3 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Updated in https://github.com/greenplum-db/gpdb/pull/3474

File: src/backend/utils/resgroup/resgroup.c [hunk 10]
```diff
@@ -1344,37 +1345,58 @@ ResGroupSlotAcquire(void)
...
+groupAcquireSlot(ResGroupData *group)
+{
+	ResGroupSlotData *slot;
 
 	/* should not been granted a slot yet */
+	Assert(selfIsAssigned());
 	Assert(!selfHasSlot());
 
+	if (selfIsAssignedDroppedGroup())
+		return NULL;
```
[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Removed in https://github.com/greenplum-db/gpdb/pull/3474

File: src/backend/utils/resgroup/resgroup.c (hunk 5 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: Updated in https://github.com/greenplum-db/gpdb/pull/3474

**Pull request closed without merging**